### PR TITLE
Add IgnoreSuffix feature

### PR DESCRIPTION
**Pull request by alonelaval**

### CODE REVIEW

File: lib/skywalking/util.lua
Comment (wu-sheng) [inline lib/skywalking/util.lua]: This line should move into the `init` stage. `split`ing every time would make performance impact.

File: test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java
Comment (wu-sheng) [inline test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java]:
```suggestion
    private String suffixEntry;

    @Before
```

File: test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java
Comment (wu-sheng) [inline test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java:70]: e2e is not checking through `Assert`. As you add a skip rule, the way to check is, after verifying traces exist, check this new endpoint's trace doesn't exist, or it doesn't exist in endpoint list through CLI.

File: lib/skywalking/util.lua
Comment (wu-sheng) [inline lib/skywalking/util.lua]: @membphis @moonming What is the fastest way to check suffix? I am feeling `find` is not a good practice.

File: lib/skywalking/tracer.lua
Comment (dmsolr) [inline lib/skywalking/tracer.lua]: Can we move it to the beginning of the method? We can return `TC.newNoOP()` directly if ignored.

File: lib/skywalking/util.lua
Comment (alonelaval) [reply lib/skywalking/util.lua]: OK

File: lib/skywalking/util.lua
Comment (membphis) [reply lib/skywalking/util.lua]: https://github.com/apache/apisix/blob/master/apisix/core/string.lua#L58

this is the better way ^_^

Comment (wu-sheng) [reply lib/skywalking/util.lua]: I think we have to copy its implementation to here, right?

File: lib/skywalking/tracer.lua
Comment (alonelaval) [reply lib/skywalking/tracer.lua]: ok

File: lib/skywalking/util.lua
Comment (alonelaval) [reply lib/skywalking/util.lua]: i got.

File: test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java
Comment (alonelaval) [reply test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java:70]: i think that‘s right，right here send request to service，suffix is ignored，then expetedData not chanaged.

File: lib/skywalking/util.lua
Comment (wu-sheng) [reply lib/skywalking/util.lua]: @alonelaval If we have to copy it, please add comments on this method about the source. We(SkyWalking and APISIX) are owned by the same foundation, so, it is OK to copy without source code License change. But still, let's make the source clear.

Comment (alonelaval) [reply lib/skywalking/util.lua]: > @alonelaval If we have to copy it, please add comments on this method about the source. We(SkyWalking and APISIX) are owned by the same foundation, so, it is OK to copy without source code License change. But still, let's make the source clear.

ok.

File: test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java
Comment (wu-sheng) [reply test/e2e/e2e-test/nginx/src/test/java/org/apache/skywalking/e2e/DataAssertITCase.java:70]: OK, we don't use the latest e2e infra tool here. We actually should apply query for this endpoint to real OAP backend, then verify it is an empty list as result.

File: lib/skywalking/tracer.lua
Comment (wu-sheng) [inline lib/skywalking/tracer.lua:38]: If you want to switch to `no_op`, please print a warning in the log, because in `==nil` case, the plugin should not be activated.

File: lib/skywalking/util.lua
Comment (wu-sheng) [inline lib/skywalking/util.lua]: Codes could change from time to time, please use perm link.

```suggestion
-- copied from: https://github.com/apache/apisix/blob/b9e36d093bbdafb6d70d332b90679ad13724c2a1/apisix/core/string.lua#L58-L67
```

File: lib/skywalking/tracer.lua
Comment (alonelaval) [reply lib/skywalking/tracer.lua:38]: ok.

File: lib/skywalking/util.lua
Comment (wu-sheng) [inline lib/skywalking/util.lua:20]: Where is `ffi` from? I can't see the dependency change.

File: test/e2e/e2e-test/nginx/pom.xml
Comment (wu-sheng) [inline test/e2e/e2e-test/nginx/pom.xml]: Why need so long time?

Comment (alonelaval) [reply test/e2e/e2e-test/nginx/pom.xml]: > Why need so long time?

in my env....

File: lib/skywalking/util.lua
Comment (alonelaval) [reply lib/skywalking/util.lua:20]: http://luajit.org/ext_ffi.html

File: lib/skywalking/tracer.lua
Comment (alonelaval) [reply lib/skywalking/tracer.lua:38]: > If you want to switch to `no_op`, please print a warning in the log, because in `==nil` case, the plugin should not be activated.

every ignored request print a warning，Will it be too much?

Comment (wu-sheng) [reply lib/skywalking/tracer.lua:38]: I am not saying `ignored` trace. You are adding logic here `serviceInstanceName == nil or serviceName == nil`. I mean this is a critical error.

File: lib/skywalking/util.lua
Comment (wu-sheng) [reply lib/skywalking/util.lua:20]: But you never declare it, and CI fails.

File: test/e2e/e2e-test/nginx/pom.xml
Comment (wu-sheng) [reply test/e2e/e2e-test/nginx/pom.xml]: Please don't change the upstream timeout, 130k seems too long.

File: lib/skywalking/util.lua
Comment (wu-sheng) [reply lib/skywalking/util.lua:20]: @membphis How should this lib be managed? Is it native in OpenRestly's Lua engine already?

**Comment by wu-sheng** [comment]
@alonelaval CI fails

https://github.com/apache/skywalking-nginx-lua/runs/4773611968?check_suite_focus=true

File: lib/skywalking/tracer.lua
Comment (alonelaval) [reply lib/skywalking/tracer.lua:38]: > I am not saying `ignored` trace. You are adding logic here `serviceInstanceName == nil or serviceName == nil`. I mean this is a critical error.
got.

File: test/e2e/e2e-test/nginx/pom.xml
Comment (alonelaval) [reply test/e2e/e2e-test/nginx/pom.xml]: OK。

File: lib/skywalking/util.lua
Comment (alonelaval) [reply lib/skywalking/util.lua:20]: > @membphis How should this lib be managed? Is it native in OpenRestly's Lua engine already?

i saw Apisix not declare too. i think native in OpenRestly.

**Comment by alonelaval** [comment]
> https://github.com/apache/skywalking-nginx-lua/runs/4773611968?check_suite_focus=true

OK，I will fix.

**Comment by tzssangglass** [comment]
hi, @alonelaval the CI installation of OpenResty includes luajit, try to use `/usr/local/openresty-debug/luajit/bin/luajit` instead of lua to test, `ffi `is a module of luajit.

**Comment by alonelaval** [comment]
> hi, @alonelaval the CI installation of OpenResty includes luajit, try to use `/usr/local/openresty-debug/luajit/bin/luajit` instead of lua to test, `ffi `is a module of luajit.

no work，in my local why luarocks installed package 'luaunit' not found.

File: lib/skywalking/util.lua
Comment (membphis) [reply lib/skywalking/util.lua:20]: > i think native in OpenRestly.

yes, you are right.

File: README.md
Comment (wu-sheng) [inline README.md]:
```suggestion
        -- set ignoreSuffix, If the operation name(HTTP URI) of the entry span includes suffixes in this set, this segment would be ignored. Multiple values should be separated by a comma(',').
```

File: examples/nginx.conf
Comment (wu-sheng) [inline examples/nginx.conf]: I think it is better to revert this? At least, we should not change this from one developer's path to another's.

File: examples/nginx.conf
Comment (wu-sheng) [inline examples/nginx.conf]: Suggest changing by following above suggestion.

File: examples/nginx.conf
Comment (wu-sheng) [inline examples/nginx.conf:114]: Please add comments here about why we need this location and how this routing rule works.

File: README.md
Comment (alonelaval) [reply README.md]: ok

File: lib/skywalking/tracer.lua
Comment (wu-sheng) [reply lib/skywalking/tracer.lua:38]: @alonelaval Don't forget this.

Comment (alonelaval) [reply lib/skywalking/tracer.lua:38]: > @alonelaval Don't forget this.

ok

**Comment by wu-sheng** [comment]
@membphis @dmsolr @tzssangglass Could you review the PR again?

**Review by membphis — APPROVED** [review]
LGTM

**Review by wu-sheng — APPROVED** [review]
Wait for more approval.

File: licenses/LICENSE-luaffi.txt
Comment (wu-sheng) [inline licenses/LICENSE-luaffi.txt]: A thing, why do we have this update? I think we don't copy the luaffi source codes.

**Review by dmsolr — APPROVED** [review]

**Review by wu-sheng — APPROVED** [review]

File: licenses/LICENSE-luaffi.txt
Comment (alonelaval) [reply licenses/LICENSE-luaffi.txt]: ok

**Pull request merged**
